Mathlib/NumberTheory/LSeries/Nonvanishing.lean

```diff
@@ -0,0 +1,237 @@
+  -- put together the various `IsBigO` statements and `norm_LFunction_product_ge_one`
+  -- to derive a contradiction
+  have H₀ : (fun _ : ℝ ↦ (1 : ℝ)) =O[𝓝[>] 0]
+      fun x ↦ LFunctionTrivChar N (1 + x) ^ 3 * LFunction χ (1 + x + I * t) ^ 4 *
+                   LFunction (χ ^ 2) (1 + x + 2 * I * t) :=
+    IsBigO.of_bound' <| eventually_nhdsWithin_of_forall
+      fun _ hx ↦ (norm_one (α := ℝ)).symm ▸ (χ.norm_LFunction_product_ge_one hx t).le
+  have H := (LFunctionTrivChar_isBigO_near_one_horizontal (N := N)).pow 3 |>.mul <|
+    (χ.LFunction_isBigO_horizontal_of_eq_zero hz₁ Hz).pow 4 |>.mul <|
+    LFunction_isBigO_horizontal _ hz₂
+  simp only [ofReal_mul, ofReal_ofNat, mul_left_comm I, ← mul_assoc, help] at H
+  -- go via absolute value to translate into a statement over `ℝ`
+  replace H := (H₀.trans H).norm_right
+  simp only [norm_eq_abs, abs_ofReal] at H
+  refine isLittleO_irrefl ?_ <| H.of_abs_right.trans_isLittleO <|
+    isLittleO_id_one.mono nhdsWithin_le_nhds
+  -- remaining goal: `∃ᶠ (x : ℝ) in 𝓝[>] 0, 1 ≠ 0`
+  simp only [ne_eq, one_ne_zero, not_false_eq_true, frequently_true_iff_neBot]
+  exact mem_closure_iff_nhdsWithin_neBot.mp <| closure_Ioi (0 : ℝ) ▸ Set.left_mem_Ici
+
+/-- If `χ` is a Dirichlet character, then `L(χ, s)` does not vanish when `s.re = 1`
+except when `χ` is trivial and `s = 1` (then `L(χ, s)` has a simple pole at `s = 1`). -/
+theorem Lfunction_ne_zero_of_re_eq_one {s : ℂ} (hs : s.re = 1) (hχs : χ ≠ 1 ∨ s ≠ 1) :
+    LFunction χ s ≠ 0 := by
+  by_cases h : χ ^ 2 = 1 ∧ s = 1
+  · exact h.2 ▸ LFunction_at_one_ne_zero_of_quadratic h.1 <| hχs.neg_resolve_right h.2
+  · have hs' : s = 1 + I * s.im := by
+      conv_lhs => rw [← re_add_im s, hs, ofReal_one, mul_comm]
+    rw [not_and_or, ← ne_eq, ← ne_eq, hs', add_right_ne_self] at h
+    replace h : χ ^ 2 ≠ 1 ∨ s.im ≠ 0 :=
+      h.casesOn .inl (fun H ↦ .inr <| by exact_mod_cast right_ne_zero_of_mul H)
+    exact hs'.symm ▸ χ.LFunction_ne_zero_of_not_quadratic_or_ne_one h
+
+/-- If `χ` is a Dirichlet character, then `L(χ, s)` does not vanish for `s.re ≥ 1`
+except when `χ` is trivial and `s = 1` (then `L(χ, s)` has a simple pole at `s = 1`). -/
+theorem Lfunction_ne_zero_of_one_le_re ⦃s : ℂ⦄ (hχs : χ ≠ 1 ∨ s ≠ 1) (hs : 1 ≤ s.re) :
+    LFunction χ s ≠ 0 :=
+  hs.eq_or_lt.casesOn (fun hs ↦ Lfunction_ne_zero_of_re_eq_one χ hs.symm hχs)
+    fun hs ↦ LFunction_eq_LSeries χ hs ▸ LSeries_ne_zero_of_one_lt_re χ hs
+
+-- Interesting special case:
+variable {χ} in
+/-- The L-function of a nontrivial Dirichlet character does not vanish at `s = 1`. -/
+theorem Lfunction_apply_one_ne_zero (hχ : χ ≠ 1) : LFunction χ 1 ≠ 0 :=
+  Lfunction_ne_zero_of_one_le_re χ (.inl hχ) <| one_re ▸ le_rfl
+
+end DirichletCharacter
+
+open DirichletCharacter in
+/-- The Riemann Zeta Function does not vanish on the closed half-plane `re s ≥ 1`. -/
+lemma riemannZeta_ne_zero_of_one_le_re ⦃s : ℂ⦄ (hz : s ≠ 1) (hz' : 1 ≤ s.re) :
```

Comment:
There is no `z` here. Also the hypothesis `hz` is redundant because of docs#riemannZeta_one_ne_zero. I know it's not a mathematically interesting statement, but it means one fewer hypothesis to verify. 

(We could even add a little argument to show non-vanishing at `s = 1` for the trivial character modulo `N > 1`, using that the factor relating the L-series of the trivial char mod N to `riemannZeta` is non-vanishing outside s = 0. But we can leave that for another PR if you prefer.)

Reply:
I had changed `z` into `s` earlier, but forgot to change the names of the assumptions, too.

I think I would like to leave removing the `χ ≠ 1` assumption in the previous lemma for a separate PR.